### DISCUSSION TITLE
Issue when i try to run mutliple line in a task

### DISCUSSION THREAD

**Discussion by Francois-lenne** [Snowflake-Labs/schemachange#370] [Q&A] [answered]
** Description**

When i try to run a a task with multiple line i have this issue : 

`snowflake.connector.errors.ProgrammingError: 001003 (42000): SQL compilation error:
syntax error line 9 at position 63 unexpected '<EOF>'.`

**To Reproduce**

This is my code (my variable ENV work i'm 100% sure) 

```
set sql_delimiter='/';
 
create or replace task TECH_{{ENV}}.TECH.TASK_ALIM_DWH_HIST
	warehouse=WH_ALIM_HP
	after TECH_{{ENV}}.TECH.TASK_ALIM_DWH
	as 

    BEGIN

        let ts_traitement timestamp_ntz := current_timestamp();
    --let  ts_traitement timestamp_ntz := to_timestamp('2024-06-01 14:00:00');

        use database DWH_{{ENV}};

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_ASSO_LOT_PROPRIETAIRE_HIST','FAITS','F_ASSO_LOT_PROPRIETAIRE','FAITS','ID_ASSO_LOT_PROPRIETAIRE');

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_BAIL_LOCATION_HIST','FAITS','F_BAIL_LOCATION','FAITS','ID_BAIL_LOCATION');
        
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_BATIMENT_HIST','FAITS','F_BATIMENT','FAITS','ID_BATIMENT');

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_CLIENT_HIST','FAITS','F_CLIENT','FAITS','ID_CLIENT');
        
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_CONTRAT_COPRO_HIST','FAITS','F_CONTRAT_COPRO','FAITS','ID_CONTRAT_COPRO');

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_FICHE_METIER_COPRO_HIST','FAITS','F_FICHE_METIER_COPRO','FAITS','ID_FICHE_METIER_COPRO');
    
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_FICHE_METIER_GESTION_HIST','FAITS','F_FICHE_METIER_GESTION','FAITS','ID_FICHE_METIER_GESTION');

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_FICHE_METIER_LOCATION_HIST','FAITS','F_FICHE_METIER_LOCATION','FAITS','ID_FICHE_METIER_LOCATION');
        
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_IMMEUBLE_HIST','FAITS','F_IMMEUBLE','FAITS','ID_IMMEUBLE');  
    
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_LOT_HIST','FAITS','F_LOT','FAITS','ID_LOT');

        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_MANDAT_GESTION_HIST','FAITS','F_MANDAT_GESTION','FAITS','ID_MANDAT_GESTION');
    
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_SATISFACTION_HIST', 'FAITS', 'F_SATISFACTION', 'FAITS', 'ID_SATISFACTION');
    
        call TECH.ALIM_DWH_TABLE_HIST(:ts_traitement, 'F_TIERS_HIST', 'FAITS', 'F_TIERS', 'FAITS', 'ID_TIERS');
    END;/

```

**Expected behavior**

In snowsight the code is running and create the task

**Schemachange (please complete the following information):**

schemachange version: **3.7.0**

**Comment by AMcNeice** [comment]
@Francois-lenne:   https://github.com/Snowflake-Labs/schemachange/issues/171#issuecomment-1641109128

I ran into the same issue the other day.

Putting the multi-line sql into a proc and then calling it in the task was the workaround that we had to do.  In your case,  could wrap all those calls into a single proc (that calls the host of them) and then call that wrapper proc in the task.

**Comment by Francois-lenne** [comment]
@AMcNeice  

Hello Andy yes i also think about this solutions i try the synthax for snowsql also but still the same error. The others solutions is probably to use the git integrations to snowflake and do an execute immediate command in a task

**Comment by ZFielden** [comment]
Just want to mention we are also seeing this issue. Hope a fix comes soon.

**Comment by sanelson** [comment]
I haven't tested this with `schemachange` yet but the following example from the bottom of the [Snowflake EXECUTE IMMEDIATE documentation](https://docs.snowflake.com/en/sql-reference/sql/execute-immediate) page helped resolve a similar issue I had running SQL scripts with the snow cli. I believe any snowflake tool using the python connector on the backend and executing a query via `execute_stream` or `execute_string` will run into this problem.

```
EXECUTE IMMEDIATE $$
DECLARE
  rs RESULTSET;
  query VARCHAR DEFAULT 'SELECT * FROM invoices WHERE price > ? AND price < ?';
  minimum_price NUMBER(12,2) DEFAULT 20.00;
  maximum_price NUMBER(12,2) DEFAULT 30.00;
BEGIN
  rs := (EXECUTE IMMEDIATE :query USING (minimum_price, maximum_price));
  RETURN TABLE(rs);
END;
$$
;
```

TLDR; wrap your code block with...

```
EXECUTE IMMEDIATE $$

....

$$
;
```

**Comment by sfc-gh-tmathew** [comment]
@Francois-lenne This is not a schemachange bug. when defining a multi-line statement appears to be an example from the snowflake docs. https://docs.snowflake.com/en/sql-reference/sql/create-task#multiple-sql-statements-using-snowsql

Based on the error message, `snowflake.connector.errors.ProgrammingError: 001003 (42000): SQL compilation error: syntax error line 9 at position 63 unexpected '<EOF>'.`, the error indicates that the SQL passed to Snowflake python connector is throwing the error and not schemachange. schemachange is not going to have a fix for this. 

The workaround to run multi-line SQL statement is to build a stored proc and then have the task call that stored proc as a single statement.

Will that suffice ?


cc: @AMcNeice @ZFielden @sanelson

**Reply by sfc-gh-tmathew** [reply, marked as answer]
Additionally, schemachange is using the python connector's execute_string() method which splits the string into statements based on semi-colon as a delimiter. The workaround until the python connector supports multi-statement tasks is either suggested above OR as suggested here in https://github.com/Snowflake-Labs/schemachange/discussions/389

Futher, will explore alternatives to address multi-line statement in a future release. For now, the workaround would be to wrap the multiline statement within EXECUTE IMMEDIATE $$ .. $$ or wrap the calls into a main proc and call that main proc. schemachange code would have replicate the statement parsing from the connector method and will need some time to reasearch. Until then, will the workaround suggested earlier or wrapping the statement with execute immediate suffice ?

**Reply by sfc-gh-tmathew** [reply]
Will be adding this pattern to the troubleshooting.md file when 4.2.0 is released.

**Reply by Francois-lenne** [reply]
hello @sfc-gh-tmathew  thanks for you return and thanks for your work 

Best regards,
François Lenne
